parsel/selector.py

```diff
@@ -267,34 +282,99 @@ def __init__(
-        self._csstranslator = _ctgroup[st]["_csstranslator"]
-        self._tostring_method = _ctgroup[st]["_tostring_method"]
+        if type not in ("html", "json", "text", "xml", None):
+            raise ValueError(f"Invalid type: {type}")
 
-        if text is not None:
-            if not isinstance(text, str):
-                msg = f"text argument should be of type str, got {text.__class__}"
-                raise TypeError(msg)
-            root = self._get_root(text, base_url)
-        elif root is None:
+        self._text = text
+
+        if text is None and root is _NOTSET:
             raise ValueError("Selector needs either text or root argument")
 
+        if text is not None and not isinstance(text, str):
+            msg = f"text argument should be of type str, got {text.__class__}"
+            raise TypeError(msg)
+
+        if text is not None:
+            if type in ("html", "xml", None):
+                self._load_lxml_root(
+                    text, type=type or "html", base_url=base_url
+                )
+            elif type == "json":
+                self.root = _load_json_or_none(text)
+                self.type = type
+            else:
+                self.root = text
+                self.type = type
+        else:
+            self.root = root
+            if type is None and isinstance(self.root, etree._Element):
+                type = "html"
+            self.type = type or "json"
+
+        self._expr = _expr
         self.namespaces = dict(self._default_namespaces)
         if namespaces is not None:
             self.namespaces.update(namespaces)
-        self.root = root
-        self._expr = _expr
+
+    def _load_lxml_root(self, text, type, base_url=None):
+        self.type = type
+        self.root = self._get_root(text, base_url)
 
     def __getstate__(self) -> Any:
         raise TypeError("can't pickle Selector objects")
 
     def _get_root(self, text: str, base_url: Optional[str] = None) -> Any:
-        return create_root_node(text, self._parser, base_url=base_url)
+        return create_root_node(
+            text,
+            _ctgroup[self.type]["_parser"],
+            base_url=base_url,
+        )
+
+    def jmespath(
+        self: _SelectorType,
+        query: str,
+        **kwargs,
+    ) -> SelectorList[_SelectorType]:
+        """
+        Find objects matching the JMESPath ``query`` and return the result as a
+        :class:`SelectorList` instance with all elements flattened. List
+        elements implement :class:`Selector` interface too.
+
+        ``query`` is a string containing the `JMESPath
+        <https://jmespath.org/>`_ query to apply.
+
+        Any additional named arguments are passed to the underlying
+        ``jmespath.search`` call, e.g.::
+
+            selector.jmespath('author.name', options=jmespath.Options(dict_cls=collections.OrderedDict))
+        """
+        if self.type == "json":
+            data = self.root
```

Comment:
What do you think about renaming type from `json` to something like `data`? json is an encoding/decoding format; I expected self.root to be a json-encoded string after seeing type="json". But actually `json` means that self.root contains some arbitraty Python object.

Reply:
To me it would be a bit like renaming "html" and "xml" to "etree". `self.type` does not represent the format after parsing, but the format before parsing, even if the input data is provided pre-parsed. And arbitrary data is not supported either, if your data has non-JSON structures chances are you are getting some error at some point, e.g. when using JMESPath.